R/calc_dist_params.R

```diff
@@ -0,0 +1,159 @@
+#' instead summary statistics of a distribution are provided, the parameters
+#' can usually be calculated from the summary statistics. The
+#' `calc_dist_params()` function computes the parameters for the gamma,
+#' lognormal and weibull distributions.
+#'
+#' This function can provide a convinient wrapper around the conversion and
+#' extraction functions it is not known which summary statistics can be used to
+#' calculate the parameters of a distribution.
+#'
+#' @details The hierarchy of methods are: conversion is prioritised if the mean
+#' and standard deviation are available as these are mostly analytical
+#' conversions (except for one of the weibull conversions), next method if
+#' possible is extraction from percentiles. This method requires a lower
+#' percentile (between(0-50]) and an upper percentile(between (50-100)). If
+#' multiple percentiles in each of these ranges is provided the lowest value is
+#' used for the calculation. The last method is the extraction using a median
+#' and range of the data.
+#'
+#' @inheritParams new_epidist
+#' @inheritParams epidist
+#' @param sample_size The sample size of the data. Only needed when falling back
+#' on using the median-range extraction calculation.
+#'
+#' @return A named vector with parameters
+#' @export
+calc_dist_params <- function(prob_dist,
+                             prob_dist_params,
+                             summary_stats,
+                             sample_size = NA) {
+
+  # check input
+  checkmate::assert_string(prob_dist)
+  checkmate::assert_list(
+    summary_stats,
+    types = c("list", "double", "null"),
+    names = "unique"
+  )
+  checkmate::assert_number(sample_size, na.ok = TRUE)
+  stopifnot(
+    "probability distribution params must be a named vector or NA" =
+      all(is.na(prob_dist_params)) ||
+      !is.null(names(prob_dist_params))
+  )
+
+  # extract mean and sd to see if conversion is possible
+  mean_sd <- c(
+    mean = summary_stats$central_tendency_spread$mean,
+    sd = summary_stats$central_tendency_spread$sd
+  )
+
+  # extract quantiles of distribution to calculate parameters as second choice
+  quantiles <- unlist(summary_stats$quantiles)
+  lower_quantiles <- quantiles[c("q_025", "q_05", "q_25", "q_50")]
+  upper_quantiles <- quantiles[c("q_75", "q_875", "q_95", "q_975")]
+
+  # extract median and range to calculate parameters as third choice
+  median_range <- c(
+    median = summary_stats$central_tendency_spread$median,
+    unlist(summary_stats$range)
+  )
+
+  if (!anyNA(mean_sd)) {
+    # make mean_sd a class with the name of the prob dist for multiple dispatch
+    prob_dist_params <- convert_params(
+      summary_stats = summary_stats,
+      prob_dist = prob_dist
+    )
+  } else if (!all(is.na(lower_quantiles)) && !all(is.na(upper_quantiles))) {
+    # use the first numeric quantile in the set to calculate the parameters
+    lower_percentile <- lower_quantiles[!is.na(lower_quantiles)][1]
```

Comment:
Do you need to check that they're symmetric? I think otherwise you might want to consider fitting with weighted least squares as the tail percentiles have more uncertainty (and similarly for fitting the ranges). I think it might be good document somewhere how the fitting is done as I for one didn't find it easy to follow in the code (and still don't fully understand it - but that might be just me).

Reply:
This is a good point. I will discuss with @adamkucharski and improve the documentation of parameter extraction and how it is used to automatically calculate parameters in `calc_dist_params()` (#102).